### DISCUSSION TITLE
How to get start with understanding firmware?

### DISCUSSION THREAD

**Discussion by cc299792458** [bitcraze/discussions#305] [Q&A]
Hi,
I have a project need to add a customized controller in the firmware. I try to understand the whole firmware. However, when I started with main.c in the folder 'src', I found I can't follow it to find where to add the controller. 
What I mean is, in the mian(), there are just some high-level functions like systemLaunch() and vTaskStartScheduler(). I want to use ctrl+click(in vscode) to follow it step by step. For example, I used ctrl+click to go into vTaskStartScheduler(), but nothing happened(That's also confusing to me).
By the way, I have found pid.c, controller_pid.c... in the folder 'modules' and I can understand them.

Could anyone teach me how to understand the stucture of the firmware? I have the basic knowledge of C and I know a little of freeRTOS. Thank you very much!

Best regards

**Comment by ataffanel** [comment]
Hi,
The `vTaskStartScheduler()` function is the FreeRTOS function that basically starts the OS. I am not sure why VsCode does not find it but it is in the [freertos souce code](https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/b0a8bd8f28d0138b5eb70e8b53da3e9d17ce8d40/tasks.c#L1954).

However, to understand the Crazyflie firmware you should likely not start reading `main.c` and you should not have to look inside FreeRTOS functions, looking at the documentation of the functions in the FreeRTOS will allow you to understand what they do.

If you really want to start from the begining, the [`systemTask`](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/system.c#L171) is the first to be launched, it initializes in sequence all the firmware modules. A module in the Crazyflie will usually have one `init()` and one `test()` function. Modules will usually spawn tasks and [wait](https://github.com/bitcraze/crazyflie-firmware/blob/e08c6a6f006be66ff4c1da72b96cda5ccba59649/src/modules/src/system.c#L330-L339) for the system to [start](https://github.com/bitcraze/crazyflie-firmware/blob/e08c6a6f006be66ff4c1da72b96cda5ccba59649/src/modules/src/system.c#L300).

Two presentation will be useful to understand the Crazyflie ecosystem and firmware architecture:
 - Introduction about the Crazyflie: https://youtu.be/jY-SLP3WWl0
 - High level firmware architecture: https://youtu.be/BXjSgCwb_bM

Finally I would suggest looking at [the documentation](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/userguides/), it documents the way some modules are intended to be used.
